### PR TITLE
Add option to log json using stdout adapter

### PR DESCRIPTION
**Pull request by indrekj**
STDOUT JSON logging allows us to transport logs to elasticsearch with context using STDOUT.

This PR also improves the logging performance.

Review commit-by-commit.

INF-1146

### CODE REVIEW

File: lib/logasm/preprocessors/whitelist.coffee [hunk 1]
```diff
@@ -1,15 +1,26 @@
 {isObject} = require('../object_helpers')
 
+isWildcardPointer = (pointer) ->
```
Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.coffee]: this flow is later replaced by a hashmap lookup https://github.com/salemove/node-logasm/pull/7/commits/6ea8b81044d6c83ce534a73c4381d3854ce71431 . probably no need to dive deep into this commit.

**Comment by ostap0207** [comment]
Did you do or can do any performance testing? Could you put some results here?

**Comment by indrekj** [comment]
Yes, I did some general performance testing. I used stdout adapter to print out 10K messages using about ~5 wildcard and 5 regular filters. Without wildcards it took ~600ms, with wildcards (using regex which was preconstructed in constructor) it took ~1.3 seconds. After switching to hashmap lookup both took ~600ms.

File: lib/logasm/preprocessors/whitelist.coffee [hunk 2]
```diff
@@ -29,11 +29,6 @@ class Whitelist
     if pointer.charAt(pointer.length - 1) == '/'
       throw Error('Pointer should not contain trailing slash')
 
-  _decodePointer: (pointer) ->
```
Comment (ostap0207) [inline lib/logasm/preprocessors/whitelist.coffee:22]: Should update doc https://github.com/salemove/node-logasm#whitelist, saying that it's a subset of JSONPointer and `~` and `/` are not supported.

**Comment by ostap0207** [comment]
One comment, otherwise 👍 . One concern though about deleting functionality, such as blacklist and `~`. I guess it OK in this case as it looks no one else is really using this lib, but in other case it wouldn't seem like a right approach. Also if we only care here about what we use or don't use maybe we should make this repo private.

**Comment by indrekj** [comment]
That's why I bumped the major version. @ostap0207 added `~` and `/` encoding back, was simple enough actually: https://github.com/salemove/node-logasm/pull/7/commits/76db97155a6201eb4b2143f1e496b5a4c2a2258b

**Review by ostap0207 — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "logasm",
   "version": "0.4.0",
-  "author": "Salemove <support@salemove.com>",
+  "author": "Salemove TechMovers <techmovers@salemove.com>",
```
Comment (deiwin) [inline package.json]: SaleMove also has a capital M :)

File: lib/logasm/adapters/stdout_adapter.coffee [hunk 4]
```diff
@@ -1,23 +1,24 @@
...
 
 class StdoutAdapter
-  constructor:(level) ->
+  constructor:(level, _serviceName, options) ->
     @logger = new winston.Logger
     @logger.add winston.transports.Console,
-      timestamp: ->
-        "#{new Date().toISOString()} ##{process.pid}"
+      json: Boolean(options.json)
+      timestamp: !Boolean(options.json) # Json adds @timestamp manually
```
Comment (deiwin) [inline lib/logasm/adapters/stdout_adapter.coffee:9]: Shouldn't it be the function it was previously when `options.json` is `false` (or falsy).

File: lib/logasm/adapters/stdout_adapter.coffee [hunk 5]
```diff
@@ -1,23 +1,24 @@
...
       level: level
       colorize: true
+      stringify: (obj) -> JSON.stringify(obj)
```
Comment (deiwin) [inline lib/logasm/adapters/stdout_adapter.coffee:12]: Kind of confused. Why is this here regardless of the `json` option?

File: lib/logasm/adapters/stdout_adapter.coffee (hunk 4 above)
Comment (deiwin) [reply lib/logasm/adapters/stdout_adapter.coffee:9]: I guess you're doing something of the sort in `buildEvent`, but why? Also there `process.pid` part isn't there anymore. Was it useful?

File: lib/logasm/object_helpers.coffee [hunk 6]
```diff
@@ -0,0 +1,5 @@
+isObject = (obj) ->
+  type = typeof obj
+  type == 'function' || type == 'object' && Boolean(obj)
```
Comment (deiwin) [inline lib/logasm/object_helpers.coffee:3]: Isn't the underscore function `obj instanceof Object` instead?

File: lib/logasm/preprocessors/whitelist.coffee (hunk 1 above)
Comment (deiwin) [reply lib/logasm/preprocessors/whitelist.coffee]: Thanks :D

**Review by deiwin — COMMENTED** [review]
> Yes, I did some general performance testing. I used stdout adapter to print out 10K messages using about ~5 wildcard and 5 regular filters. Without wildcards it took ~600ms, with wildcards (using regex which was preconstructed in constructor) it took ~1.3 seconds. After switching to hashmap lookup both took ~600ms.

Can you include a similar performance test in the project so that this would be automatically run for all PRs?

File: lib/logasm/adapters/stdout_adapter.coffee (hunk 4 above)
Comment (indrekj) [reply lib/logasm/adapters/stdout_adapter.coffee:9]: the default timestamp is okay. Process pid is not useful at all with kubernetes (all processes are pid 0). I would say it wasn't even useful with the previous deployment.

File: lib/logasm/adapters/stdout_adapter.coffee (hunk 5 above)
Comment (indrekj) [reply lib/logasm/adapters/stdout_adapter.coffee:12]: `json: true` has `JSON.stringify(obj, null, 2)` which means it creates multiline json objects. We need one-line logs.

File: lib/logasm/object_helpers.coffee (hunk 6 above)
Comment (indrekj) [reply lib/logasm/object_helpers.coffee:3]: nope https://github.com/jashkenas/underscore/blob/master/underscore.js#L1313-L1315

**Comment by indrekj** [comment]
@deiwin Added https://github.com/salemove/node-logasm/pull/7/commits/f8940606817bf74628dcb74ad167e9ac009f91cb

Same test on master: `AssertionError: expected 3670 to be at most 100`.

**Review by deiwin — APPROVED** [review]

**Comment by indrekj** [comment]
!merge
